scripts/check_regression.sh

```diff
@@ -0,0 +1,81 @@
+#!/bin/bash
+
+set -eu -o pipefail
+
+# Check if two arguments are provided (paths to past.json and current.json)
+if [ $# -ne 2 ]; then
+    echo "Usage: $0 <path_to_past.json> <path_to_current.json>"
+    exit 1
+fi
+
+# Extract the file paths from command-line arguments
+past_json_path="$1"
+current_json_path="$2"
+
+# Read the contents of past.json and current.json into variables
+past_data=$(cat "$past_json_path")
+current_data=$(cat "$current_json_path")
+
+# Function to compare P90 values for a given statistic
+compare_stat_p90() {
+    local past_value="$1"
+    local current_value="$2"
+    local stat_name="$3"
+
+    # Calculate 110% of the past value
+    local threshold=$(calculate_threshold "$past_value")
+
+    # Compare the current value with the threshold
+    if (( $(awk 'BEGIN {print ("'"$current_value"'" > "'"$threshold"'")}') )); then
+        echo "ERROR: $stat_name - Current P90 value ($current_value) exceeds the 110% threshold ($threshold) of the past P90 value ($past_value)"
+        return 1
+    fi
+
+    return 0
+}
+
+calculate_threshold() {
+    local past_value="$1"
+    awk -v past="$past_value" 'BEGIN { print past * 1.1 }'
+}
+
+# Loop through each object in past.json and compare P90 values with current.json for all statistics
+compare_p90_values() {
+    local past_json="$1"
+    local current_json="$2"
+
+    local test_names=$(echo "$past_json" | jq -r '.benchmarkTests[].testName')
+
+    # Use a flag to indicate if any regression has been detected
+    local regression_detected=0
+
+    for test_name in $test_names; do
+        echo "Checking for regression in '$test_name'"
+        for stat_name in "fullRunStats" "pullStats" "lazyTaskStats" "localTaskStats"; do
+            local past_p90=$(echo "$past_json" | jq -r --arg test "$test_name" '.benchmarkTests[] | select(.testName == $test) | .'"$stat_name"'.pct90')
+            local current_p90=$(echo "$current_json" | jq -r --arg test "$test_name" '.benchmarkTests[] | select(.testName == $test) | .'"$stat_name"'.pct90')
+
+            # Call the compare_stat_p90 function
+            compare_stat_p90 "$past_p90" "$current_p90" "$stat_name" || regression_detected=1
+        done
+    done
+
+    # Check if any regression has been detected and return the appropriate exit code
+    return $regression_detected
+}
+
+# ... (remaining code)
+
+# Call compare_p90_values and store the exit code in a variable
+compare_p90_values "$past_data" "$current_data"
+exit_code=$?
```

Comment:
Does this work with `set -e`? Won't it exit when `compare_p90_values` returns 1?

Reply:
It should work since its the first part of `||`

re: https://www.gnu.org/software/bash/manual/html_node/The-Set-Builtin.html
> The shell does not exit if the command that fails is part of the command list immediately following a while or until keyword, part of the test in an if statement, part of any command executed in a && or || list except the command following the final && or ||

Reply:
> The set -e option instructs bash to immediately exit if any command [1] has a non-zero exit status. You wouldn't want to set this for your command-line shell, but in a script it's massively helpful. In all widely used general-purpose programming languages, an unhandled runtime error
whether that's a thrown exception in Java, or a segmentation fault in C, or a syntax error in Python - immediately halts execution of the program; subsequent lines are not executed.

My understanding is that, If a function returns a non-zero value, it is considered a regular function return value, and the script will continue executing without terminating due to set -e.

I tested it out by running the workflow in my fork (console logs below), it seems to be working as expected and not exiting after the first regression is identified.

```
Checking for regression in 'SociFullECR-public-ffmpeg'
Checking for regression in 'SociFullECR-public-tensorflow'
ERROR: fullRunStats - Current P90 value (20.328798902) exceeds the 110% threshold (13.5617) of the past P90 value (12.328798902)
ERROR: localTaskStats - Current P90 value (2.526) exceeds the 110% threshold (1.6786) of the past P90 value (1.526)
Checking for regression in 'SociFullECR-public-tensorflow_gpu'
ERROR: pullStats - Current P90 value (39.567) exceeds the 110% threshold (38.0237) of the past P90 value (34.567)
ERROR: localTaskStats - Current P90 value (2.6) exceeds the 110% threshold (1.7787) of the past P90 value (1.617)
Checking for regression in 'SociFullECR-public-node'
Checking for regression in 'SociFullECR-public-busybox'
Checking for regression in 'SociFullECR-public-mongo'
Checking for regression in 'SociFullECR-public-rabbitmq'
Checking for regression in 'SociFullECR-public-redis'
Comparison failed. Current P90 values exceed 110% of the corresponding past values.
```